torch/distributed/algorithms/model_averaging/utils.py

```diff
@@ -19,18 +19,45 @@ def average_parameters(
     if dist._rank_not_in_group(group_to_use):
         return
 
+    if isinstance(params, types.GeneratorType):
+        # compatible with model.parameters() input
+        flat_params_all_reduce(params, group_to_use)
+    elif isinstance(params, list):
+        param_groups = params
+        # support optim.param_group input
+        if not comm_memory_efficient:
+            params_list = get_param_list(param_groups)
+            flat_params_all_reduce(params_list, group_to_use)
+        else:
+            for param_group in param_groups:
+                for param in param_group["params"]:
+                    if param.grad is None:
+                        continue
+                    flat_params_all_reduce([param], group_to_use)
+
+
+def flat_params_all_reduce(params: Union[types.GeneratorType, List], process_group: dist.ProcessGroup):
     params_it1, params_it2 = itertools.tee(params)
     # If the input parameters have different data types,
     # packing these parameters will trigger an implicit type up-casting.
     # The original parameter data types will be restored during the subsequent unpacking.
-    flat_params = torch.cat([p.data.view(-1) for p in params_it1])
-    flat_params /= dist.get_world_size(group_to_use)
+    flat_params = torch.cat([p.data.reshape(-1) for p in params_it1])
+    flat_params /= dist.get_world_size(process_group)
```

Comment:
Nit: Personally prefer `group_to_use` to the input `process_group`.

Understand that they are equivalent, but when the input `process_group` is `None`, `group_to_use` will be `WORLD`, and it will be more explicit and readable. Otherwise, readers may need to to check whether `None` and `WORLD` will ahve the same effect on `get_world_size` and `all_reduce`. At least they don't have the same effect on `_rank_not_in_group`.

Reply:
agree, i will fix later

Reply:
fix already. 
torch/distributed/algorithms/model_averaging/utils.py 
only change 2 parts.
1. p.data.view   -> p.data.reshape
2. Iterator[torch.nn.Parameter]   -> List[torch.nn.Parameter]